### ISSUE TITLE
Cannot list directories; "No such file or directory"

### ISSUE THREAD

**Issue by hills** [alperakcan/fuse-ext2#12]
Like many other people, I am trying to recover data from a NAS; in this case a Seagate Central which uses a 64kb block size that is not compatible with my x86 system.

I can mount my partition using fuse-ext2, but once mounted I'm unable to list any directory (all command run as root)

>  # cd /mnt/central
>  # ls
>  ls: cannot open directory .: No such file or directory

Is this a fuse-ext2 bug? Or some consequence of the file system or inconsistency, perhaps. Although I've not been able to find any reports of this, including people recovering from the same devices without this problem.

It appears I could perhaps read files or directories if I know their name. I found I am able to cd to a directory called 'Public' (or other directories I know exist from this post [1]) but not others:

[1] http://ubuntuforums.org/showthread.php?t=2199589&page=3&p=12904552#post12904552

>  # cd test
>  bash: cd: test: No such file or directory
>  #stat Public
>    File: 'Public'
>    Size: 65536           Blocks: 128        IO Block: 65536  directory
>  Device: 16h/22d Inode: 2           Links: 33
>  Access: (2775/drwxrwsr-x)  Uid: (65534/ UNKNOWN)   Gid: (65534/ UNKNOWN)
>  Access: 2015-06-14 11:24:37.000000000 +0100
>  Modify: 2015-07-30 22:45:37.000000000 +0100
>  Change: 2015-06-14 11:23:56.000000000 +0100
>   Birth: -
>  # cd Public
>  # ls
>  ls: cannot open directory .: No such file or directory

I've tried the latest Git master (213f776e4), as well as release 0.0.9. A full transcript of the debug for my actions above is as follows:

>  # /opt/fuse-ext2/bin/fuse-ext2 -v /dev/vg1/lv1 /mnt/central/  -o debug,ro
FUSE library version: 2.8.5
nullpath_ok: 0
unique: 1, opcode: INIT (26), nodeid: 0, insize: 56
INIT: 7.23
flags=0x0003f7fb
max_readahead=0x00020000
   INIT: 7.12
   flags=0x00000011
   max_readahead=0x00020000
   max_write=0x00020000
   unique: 1, success, outsize: 40
unique: 2, opcode: GETATTR (3), nodeid: 1, insize: 56
getattr /
   unique: 2, success, outsize: 120
unique: 3, opcode: GETATTR (3), nodeid: 1, insize: 56
getattr /
   unique: 3, success, outsize: 120
unique: 4, opcode: OPENDIR (27), nodeid: 1, insize: 48
opendir flags: 0x18800 /
   unique: 4, error: -2 (No such file or directory), outsize: 16
unique: 5, opcode: GETATTR (3), nodeid: 1, insize: 56
getattr /
   unique: 5, success, outsize: 120
unique: 6, opcode: LOOKUP (1), nodeid: 1, insize: 45
LOOKUP /test
getattr /test
   unique: 6, error: -2 (No such file or directory), outsize: 16
unique: 7, opcode: LOOKUP (1), nodeid: 1, insize: 45
LOOKUP /test
getattr /test
   unique: 7, error: -2 (No such file or directory), outsize: 16
unique: 8, opcode: LOOKUP (1), nodeid: 1, insize: 45
LOOKUP /test
getattr /test
   unique: 8, error: -2 (No such file or directory), outsize: 16
unique: 9, opcode: GETATTR (3), nodeid: 1, insize: 56
getattr /
   unique: 9, success, outsize: 120
unique: 10, opcode: LOOKUP (1), nodeid: 1, insize: 47
LOOKUP /Public
getattr /Public
   NODEID: 2
   unique: 10, success, outsize: 144
unique: 11, opcode: GETATTR (3), nodeid: 1, insize: 56
getattr /
   unique: 11, success, outsize: 120
unique: 12, opcode: LOOKUP (1), nodeid: 1, insize: 47
LOOKUP /Public
getattr /Public
   NODEID: 2
   unique: 12, success, outsize: 144
unique: 13, opcode: GETATTR (3), nodeid: 2, insize: 56
getattr /Public
   unique: 13, success, outsize: 120
unique: 14, opcode: OPENDIR (27), nodeid: 2, insize: 48
opendir flags: 0x18800 /Public
   unique: 14, error: -2 (No such file or directory), outsize: 16

Thanks in advance
Mark

**Comment by madebr** [comment]
I can reproduce this error on Fedora 22.

Create an empty filesystem:
```
[maarten@localhost fuse_ext2_test]$ dd if=/dev/zero of=test.img bs=512 count=10000
10000+0 records in
10000+0 records out
5120000 bytes (5.1 MB) copied, 0.0222437 s, 230 MB/s
[maarten@localhost fuse_ext2_test]$ mkfs.ext2 test.img 
mke2fs 1.42.12 (29-Aug-2014)
Discarding device blocks: done                            
Creating filesystem with 5000 1k blocks and 1256 inodes

Allocating group tables: done                            
Writing inode tables: done                            
Writing superblocks and filesystem accounting information: done
```
Let the mount point be mnt:
```
[maarten@localhost fuse_ext2_test]$ mkdir mnt
```
Mount the image read-only using fuse-ext2.
```
[maarten@localhost fuse_ext2_test]$ ./fuse-ext2 test.img mnt -o ro
[maarten@localhost fuse_ext2_test]$ ls mnt
ls: cannot open directory mnt: No such file or directory
[maarten@localhost fuse_ext2_test]$ fusermount -u mnt
```
Mount the image using the kernel.
```
[maarten@localhost fuse_ext2_test]$ sudo mount -o loop test.img mnt
[maarten@localhost fuse_ext2_test]$ ls mnt
lost+found
[maarten@localhost fuse_ext2_test]$ sudo umount mnt
```

**Comment by hills** [comment]
Thanks. Your information gave me enough of a lead, and a simpler test case too.

It seems the problem isn't with *reading* the directory, but with this snippet of code from `op_open.c`:

```
ext2_file_t do_open (ext2_filsys e2fs, const char *path, int flags)
{
        // code to read the directory here
        [...]

        inode.i_mtime = e2fs->now ? e2fs->now : time(NULL);
        rt = do_writeinode(e2fs, ino, &inode);
        if (rt) {
                debugf("do_writeinode(%s, &ino, &inode); failed", path);
                return NULL;
        }
```

At a glance, this looks all wrong to me. Aside from the printed debug message having the wrong suffix, I don't understand why we would attempt to write back the inode with a new mtime (modify time) on open. Is this actually a slightly ham-fisted and typo'd attempt to update the *atime* (access time)? Because it also doesn't change its behaviour if the filesystem is mounted read-only either.

In any case, removing this code allowed me to list directory contents and make a start on the data recovery I need.

But something doesn't seem right with this block of code, and I'm surprised to be the first person to hit this, too.

The code in question appears in commit 0528a7c4 (ie. between v0.0.8 and v0.0.9), but the patch description only says "pjdfstest fixes" and doesn't offer any more information why this was added.

**Comment by alperakcan** [comment]
hi there,

hills you are right, i think it is some kind of typo. it should be atime. and time should be modified only if read-write mounted.

i think i should change mtime in op_write, atime in op_open

**Comment by alperakcan** [comment]
fixed with commit f7fa0d962d8c605423cb865793273965b0039ba8
